controllers/devworkspace/solver/che_routing.go

```diff
@@ -208,212 +270,297 @@ func isSecureScheme(scheme string) bool {
+		mdls[dwId+"-auth"] = traefikConfigMiddleware{
+			ForwardAuth: &traefikConfigForwardAuth{
+				Address: "http://127.0.0.1:8089?namespace=" + dwNamespace,
+			},
 		}
 
-		ports := multihostPorts
-		if gatewayEnabled && e.Attributes.GetString(urlRewriteSupportedEndpointAttributeName, nil) == "true" {
-			ports = singlehostPorts
+		mdls[dwId+"-header"] = traefikConfigMiddleware{
+			Plugin: &traefikPlugin{
+				HeaderRewrite: &traefikPluginHeaderRewrite{
+					From:   "X-Forwarded-Access-Token",
+					To:     "Authorization",
+					Prefix: "Bearer ",
+				},
+			},
 		}
+	}
 
-		if ports[i] == nil {
-			ports[i] = map[string]portMappingValue{}
+	if len(cfg.HTTP.Routers) > 0 {
+		contents, err := yaml.Marshal(cfg)
+		if err != nil {
+			logger.Error(err, "can't serialize traefik config")
+			return nil
 		}
 
-		if _, ok := ports[i][name]; !ok {
-			ports[i][name] = portMappingValue{
-				order:          *order,
-				endpointScheme: determineEndpointScheme(gatewayEnabled, e),
-			}
-			*order = *order + 1
+		configMap := &corev1.ConfigMap{
+			ObjectMeta: v1.ObjectMeta{
+				Name:      defaults.GetGatewayWorkspaceConfigMapName(dwId),
+				Namespace: cheCluster.Namespace,
+				Labels:    cmLabels,
+				Annotations: map[string]string{
+					defaults.ConfigAnnotationDevWorkspaceRoutingName:      routing.Name,
+					defaults.ConfigAnnotationDevWorkspaceRoutingNamespace: routing.Namespace,
+				},
+			},
+			Data: map[string]string{},
 		}
+		configMap.Data[dwId+".yml"] = string(contents)
+		return configMap
 	}
-
-	return
+	return nil
 }
 
-func addToTraefikConfig(namespace string, workspaceID string, machineName string, portMapping portMapping, cfg *traefikConfig) {
-	rtrs := cfg.HTTP.Routers
-	srvcs := cfg.HTTP.Services
-	mdls := cfg.HTTP.Middlewares
-
-	for port, names := range portMapping {
-		for endpointName := range names {
-			name := getEndpointExposingObjectName(machineName, workspaceID, port, endpointName)
-			var prefix string
-			var serviceURL string
-
-			prefix = getPublicURLPrefix(workspaceID, machineName, port, endpointName)
-			serviceURL = getServiceURL(port, workspaceID, namespace)
+func addEndpointToTraefikConfig(componentName string, e dw.Endpoint, cfg *traefikConfig, cheCluster *v2alpha1.CheCluster, routing *dwo.DevWorkspaceRouting) {
+	var routeName string
+	if e.Attributes.GetString(uniqueEndpointAttributeName, nil) == "true" {
+		// if endpoint is unique, we're exposing on /componentName/<endpoint-name>
+		routeName = e.Name
+	} else {
+		// if endpoint is NOT unique, we're exposing on /componentName/<port-number>
+		routeName = strconv.Itoa(e.TargetPort)
+	}
+	prefix := fmt.Sprintf("/%s/%s", componentName, routeName)
+	rulePrefix := fmt.Sprintf("PathPrefix(`%s`)", prefix)
 
-			rtrs[name] = traefikConfigRouter{
-				Rule:        fmt.Sprintf("PathPrefix(`%s`)", prefix),
-				Service:     name,
-				Middlewares: calculateMiddlewares(name),
-				Priority:    100,
-			}
+	// skip if exact same route is already exposed
```

Comment:
Under what circumstances does this happen? Can we eliminate the possibility sooner?

Reply:
multiple non-unique subpath endpoints. There's test for it https://github.com/eclipse-che/che-operator/blob/main/controllers/devworkspace/solver/che_routing_test.go#L152

Reply:
In theory, yes we can eliminate sooner and it is how it's currently implemented. However, I think this is much more straightforward. It was really hard to do even small change without breaking it completely, because endpoints were organized to complicated maps by various properties with custom structs etc. So small change to one of those things lead to completely different result. This just accepts the original endpoint from devworkspace-controller and expose it as needed (or not if it is already exposed). That's why I ended up with doing this bigger refactoring.